server/src-lib/Hasura/GraphQL/Execute/Inline.hs

```diff
@@ -0,0 +1,176 @@
+further processing, since we catch unbound fragments and recursive fragment
+definitions early in the pipeline, so parsing does not have to worry about it.
+In that sense, fragment inlining is similar to the variable resolution pass
+performed by "Hasura.GraphQL.Execute.Resolve", but for fragment definitions
+rather than variables. -}
+module Hasura.GraphQL.Execute.Inline
+  ( inlineSelectionSet
+  ) where
+
+import           Hasura.Prelude
+
+import qualified Data.HashMap.Strict.Extended  as Map
+import qualified Data.HashSet                  as Set
+import qualified Data.List                     as L
+import qualified Data.Text                     as T
+
+import           Control.Lens
+import           Language.GraphQL.Draft.Syntax
+
+import           Hasura.RQL.Types.Error
+import           Hasura.Server.Utils
+import           Hasura.SQL.Types
+
+-- | Internal bookkeeping used during inlining.
+data InlineEnv = InlineEnv
+  { _ieFragmentDefinitions :: HashMap Name FragmentDefinition
+  -- ^ All known fragment definitions.
+  , _ieFragmentStack       :: [Name]
+  -- ^ Fragments we’re currently inlining higher up in the call stack, used to
+  -- detect fragment cycles.
+  }
+
+-- | Internal bookkeeping used during inlining.
+newtype InlineState var = InlineState
+  { _isFragmentCache :: HashMap Name (InlineFragment NoFragments var)
+  -- ^ A cache of fragment definitions we’ve already inlined, so we don’t need
+  -- to inline them again.
+  }
+
+$(makeLensesFor [("_ieFragmentStack", "ieFragmentStack")] ''InlineEnv)
+$(makeLenses ''InlineState)
+
+type MonadInline var m =
+  ( MonadError QErr m
+  , MonadReader InlineEnv m
+  , MonadState (InlineState var) m
+  )
+
+-- | Inlines all fragment spreads in a 'SelectionSet'; see the module
+-- documentation for "Hasura.GraphQL.Execute.Inline" for details.
+inlineSelectionSet
+  :: (MonadError QErr m, Foldable t)
+  => t FragmentDefinition
+  -> SelectionSet FragmentSpread var
+  -> m (SelectionSet NoFragments var)
+inlineSelectionSet fragmentDefinitions selectionSet = do
+  let fragmentDefinitionMap = Map.groupOnNE _fdName fragmentDefinitions
+  uniqueFragmentDefinitions <- flip Map.traverseWithKey fragmentDefinitionMap
+    \fragmentName fragmentDefinitions' ->
+      case fragmentDefinitions' of
+        a :| [] -> return a
+        _       -> throw400 ParseFailed $ "multiple definitions for fragment " <>> fragmentName
+  let usedFragmentNames = Set.fromList $ fragmentsInSelectionSet selectionSet
+      definedFragmentNames = Set.fromList $ Map.keys uniqueFragmentDefinitions
+      isFragmentValidationEnabled = False
+  when (isFragmentValidationEnabled && (usedFragmentNames /= definedFragmentNames)) $
+    throw400 ValidationFailed $
+    ("following fragment(s) have been defined, but have not been used in the query - "
+    <> (T.concat $ L.intersperse ", "
+       $ map unName $ Set.toList $
+       Set.difference definedFragmentNames usedFragmentNames))
```

Comment:
I’m not fond of having totally dead code like this. @abooij, do you have any bigger plans for this (like adding a flag to enable it), or should we remove it for now? (We can always put it back later.)

Reply:
@lexi-lambda Actually, @codingkarthik wrote this code. After having written it, @nicuveo and I made the judgment call that it was perhaps unwise to introduce additional incompatibilities with master as there is no corresponding check there. Hence, it was decided to disable this code for now, and re-enable it later. So the intention is to enable this globally, without a flag, at a later stage (perhaps after the first release that includes PDV, say).

I think it would be reasonable to involve some of the above named developers in the decision whether to remove this code.

Reply:
On one hand I tend to be on the side of deleting the code; it's in the git history where it belongs anyway. On the other, keeping it in avoids the risk of duplicating work; and having it compiled rather than commented means it isn't subject to bit rot: we won't go back to a comment that doesn't compile anymore.
Ideally, the best solution here would be to tie it to a flag that can never be set to True for now; meaning that whenever we want to activate this feature we don't have to change this part of the code. But we have no existing structure for feature flags that I know of?

Reply:
There's a variables usage validation also which is disabled like we have disabled the fragment validation here. I agree with @nicuveo , keeping the code in, will avoid duplicating work. But, the current code is also not very neat. I would personally like this be controlled by a feature flag as mentioned above, but if it's too much work at this point, given that we're close to merge PDV, we can delete this code and maybe create an issue, referencing this part of code?

Reply:
Yes, I don’t think we currently have any feature flags… though I could be wrong. But it would basically just involve plumbing the information into the right place. My preference would be to do that if we want to keep the flag, since if we decide to make it the default in the future, we can change the flag to opt out of the strict checking instead of opting in (as a compatibility option: eventually we could remove the flag altogether).